CoreLib/SharedPtr.h

```diff
@@ -0,0 +1,490 @@
+﻿#pragma once
+
+
+#include "CoreLibConcepts.h"
+
+#include <Core.h>
+
+#include <stdint.h>
+
+namespace Fuego
+{
+    template<class T, class Deleter> requires IsDefaultCompatibleDeleter<T, Deleter>
+    class SharedPtr;
+
+    template<class T, class Deleter = std::default_delete<T>> requires IsDefaultCompatibleDeleter<T, Deleter>
+    class FUEGO_API WeakPtr
+    {
+    public:
+
+        WeakPtr() noexcept = default;
+        WeakPtr(const WeakPtr<T, Deleter>& other) noexcept;
+        WeakPtr(WeakPtr<T, Deleter>&& other) noexcept;
+
+        WeakPtr<T, Deleter>& operator=(const WeakPtr<T, Deleter>& other) noexcept;
+        WeakPtr<T, Deleter>& operator=(WeakPtr<T, Deleter>&& other) noexcept;
+
+        ~WeakPtr();
+
+        WeakPtr(const  SharedPtr<T, Deleter>& shared) noexcept;
+        WeakPtr<T, Deleter>& operator=(const SharedPtr<T, Deleter>& shared) noexcept;
+
+        SharedPtr<T, Deleter> Lock() const noexcept;
+
+        bool Expired() const noexcept;
+        void Release() noexcept;
+        void Swap(WeakPtr<T, Deleter>& other) noexcept;
+
+    private:
+        typename SharedPtr<T, Deleter>::_ControlBlock* _control_block = nullptr;
+    };
+
+    template<class T, class Deleter = std::default_delete<T>> requires IsDefaultCompatibleDeleter<T, Deleter>
+    class FUEGO_API SharedPtr
+    {
+    public:
+        SharedPtr() noexcept = default;
+
+        explicit SharedPtr(std::remove_extent_t<T>* ptr);
+        SharedPtr(const SharedPtr<T, Deleter>& other) noexcept;
+        SharedPtr(SharedPtr<T, Deleter>&& other) noexcept;
+
+        SharedPtr<T, Deleter>& operator=(const SharedPtr<T, Deleter>& other) noexcept;
+        SharedPtr<T, Deleter>& operator=(SharedPtr<T, Deleter>&& other) noexcept;
+
+        ~SharedPtr();
+
+        std::remove_extent_t<T> const * Get() const noexcept;
+        std::remove_extent_t<T>* Get() noexcept;
+
+        std::remove_extent_t<T>  const & operator*() const noexcept;
+        std::remove_extent_t<T>& operator*() noexcept;
+
+        std::remove_extent_t<T>  const * operator->() const noexcept;
+        std::remove_extent_t<T>* operator->() noexcept;
+
+        uint32_t UseCount() const noexcept;
+        explicit operator bool() const noexcept;
+
+        void Reset(std::remove_extent_t<T>* ptr = nullptr);
+        void Release();
+
+        void Swap(SharedPtr<T, Deleter>& other) noexcept;
+
+        template<class U, class DeleterU> requires IsDefaultCompatibleDeleter<U, DeleterU>
+        friend class WeakPtr;
+
+        template<class U>
+        bool operator==(std::remove_extent_t<U>* other) const noexcept;
```

Comment:
https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_cmp

Does it make sense to follow the pre C++20 API here?

Reply:
Good point